src/compute/src/render.rs

```diff
@@ -223,7 +225,24 @@ pub fn build_compute_dataflow<A: Allocate>(
             // Import declared sources into the rendering context.
             for (source_id, (source, _monotonic)) in dataflow.source_imports.iter() {
                 region.region_named(&format!("Source({:?})", source_id), |inner| {
-                    let mut mfp = source.arguments.operators.clone().map(|ops| {
+                    let mut read_schema = None;
+                    let mut mfp = source.arguments.operators.clone().map(|mut ops| {
+                        // If enabled, we read from Persist with a `RelationDesc` that
+                        // omits uneeded columns.
+                        if apply_demands {
+                            let demands = ops.demand();
+                            let new_desc =
+                                source.storage_metadata.relation_desc.apply_demand(&demands);
+                            let new_arity = demands.len();
+                            let remap = demands
+                                .into_iter()
+                                .enumerate()
+                                .map(|(new, old)| (old, new))
+                                .collect();
+                            ops.permute(remap, new_arity);
+                            read_schema = Some(new_desc);
+                        }
```

Comment:
FWIW, some things I think might make this a little cleaner:
- The `demand` + `permute` could be a new helper on the MFP itself... `pre_project` or something? May help clarify intent.
- This logic could live inside the `persist_source` itself... though I guess you'd need to move the `MfpPlan::create_from` in there too. My feeling is that (rollout aside) you'd never not want to apply this projection, so better to not make it the caller's responisibility.

Neither suggestion blocking!

Reply:
I played around with adding a `demand` + `permute` helper on the MFP itself, but couldn't get something that I really liked. Mainly because applying the demands on a `RelationDesc` returns a new `RelationDesc` instead of modifying in-place, which I think requires the helper to be able to return the new `RelationDesc`. It could also be a case of morning brain, if you have something specific you're thinking of here let me know, happy to circle back on it!